lib/generate_patients.rb

```diff
@@ -0,0 +1,128 @@
+          race = generate_loaded_datatype('QDM::PatientCharacteristicRace')
+          ethnicity = generate_loaded_datatype('QDM::PatientCharacteristicEthnicity')
+          birthdate = generate_loaded_datatype('QDM::PatientCharacteristicBirthdate')
+          qdm_patient.dataElements.push(sex)
+          qdm_patient.dataElements.push(race)
+          qdm_patient.dataElements.push(ethnicity)
+          qdm_patient.dataElements.push(birthdate)
+          cqm_patient.qdmPatient = qdm_patient
+        end
+
+        data_element = generate_loaded_datatype(type)
+        qdm_patient.dataElements.push(data_element)
+        # if type is negatable, add a negated version to the patient
+        if data_element.fields.keys.include? 'negationRationale'
+          negated_data_element = generate_loaded_datatype(type, true)
+          qdm_patient.dataElements.push(negated_data_element)
+        end
+        patients.push(cqm_patient)
+      end
+      patients
+    end
+
+    def self.get_datatypes(modelinfo_file)
+      modelinfo = File.open(modelinfo_file) { |f| Nokogiri::XML(f) }
+      datatypes = []
+      # Loop through each typeInfo node (each of these is a QDM datatype)
+      modelinfo.xpath('//ns4:typeInfo').each do |type|
+        # Grab the name of this QDM datatype
+        datatype_name = type.attributes['name'].value.split('.').last
+        exclusion_array = %w[Component Id Patient ResultComponent FacilityLocation]
+        # Store datatype and its attributes (reject irrelevant datatypes)
+        next if datatype_name.include?('Negative') || datatype_name.include?('Positive') ||
+                datatype_name.include?('QDMBaseType') || (exclusion_array.include? datatype_name)
+        datatypes.push(datatype_name)
+      end
+      datatypes
+    end
+
+    def self.generate_loaded_datatype(data_element_type, negate_data_element = false)
+      data_element = QDM.const_get(data_element_type).new
+      fields = data_element.typed_attributes.keys
+      fields.each do |field_name|
+        # Ignore these fields, they are used by mongoose
+        next if %w[_id _type].include? field_name
+        if !data_element[field_name] || data_element[field_name] == []
+          populate_fields(field_name, data_element, negate_data_element)
+        end
+      end
+      data_element
+    end
+
+    def self.populate_fields(field_name, data_element, negate_data_element)
+      # There are certain fields that we want to populate manually
+      if field_name == 'description'
+        # Skip the setting of patient description for now
+      elsif field_name == 'negationRationale'
+        data_element[field_name] = QDM::BaseTypeGeneration.generate_code_field if negate_data_element
+      elsif field_name == 'components'
+        data_element[field_name] = [QDM::BaseTypeGeneration.generate_component]
+      elsif field_name == 'result'
+        # TODO: Result can be MANY Integer, Decimal, Code, Quantity or Ratio randomize this
```

Comment:
Is there a planned task in the future to address these TODOs? If not, perhaps we should remove all of the TODOs and have one more general todo of randomizing the generated data.

Reply:
https://jira.mitre.org/browse/CYPRESS-432